packages/tailwindcss/src/index.ts

```diff
@@ -188,9 +191,307 @@ async function parseCss(
+                          else if (
+                            type === 'percentage' &&
+                            !isPositiveInteger(value.slice(0, -1))
+                          ) {
+                            continue
+                          }
+
+                          resolvedValueType = type
+                          resolvedValueFn = true
+                          replaceWith(ValueParser.parse(value))
+                          return ValueParser.ValueWalkAction.Skip
+                        }
+                      }
+
+                      // Arbitrary value, e.g.: `value([integer])`
+                      else if (
+                        candidate.value?.kind === 'arbitrary' &&
+                        arg.kind === 'word' &&
+                        arg.value[0] === '[' &&
+                        arg.value[arg.value.length - 1] === ']'
+                      ) {
+                        let dataType = arg.value.slice(1, -1)
+
+                        // Allow any data type, e.g.: `value([*])`
+                        if (dataType === '*') {
+                          resolvedValueFn = true
+                          replaceWith(ValueParser.parse(candidate.value.value))
+                          return ValueParser.ValueWalkAction.Skip
+                        }
+
+                        // The forced arbitrary value hint must match the
+                        // expected data type.
+                        //
+                        // ```css
+                        // @utility tab-* {
+                        //   tab-size: value([integer]);
+                        // }
+                        // ```
+                        //
+                        // Given a candidate like `tab-(color:var(--my-value))`,
+                        // should not match because `color` and `integer` don't
+                        // match.
+                        if (candidate.value.dataType && candidate.value.dataType !== dataType) {
+                          continue
+                        }
+
+                        let value = candidate.value.value
+                        let type =
+                          candidate.value.dataType ?? inferDataType(value, [dataType as any])
+
+                        if (type !== null) {
+                          resolvedValueFn = true
+                          replaceWith(ValueParser.parse(value))
+                          return ValueParser.ValueWalkAction.Skip
+                        }
+                      }
+                    }
+
+                    // Drop the declaration in case we couldn't resolve the value
+                    usedValueFn ||= false
+                    replaceDeclarationWith([])
+                    return ValueParser.ValueWalkAction.Stop
+                  }
+
+                  // Modifier function, e.g.: `modifier(integer)`
+                  else if (valueNode.value === 'modifier') {
```

Comment:
There's quite some repetition between the modifier and the value function, maybe we can avoid some code duplication here by adding a more generic resolver function?

Reply:
Merged the codepaths. There was some handling that isn't exactly the same (around `ratio` and `fraction`, but the differences are not too bad).